google/cloud/opentelemetry/configure_basic_tracing.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/opentelemetry/configure_basic_tracing.h"
+#include "google/cloud/opentelemetry/trace_exporter.h"
+#include <opentelemetry/sdk/trace/batch_span_processor.h>
+#include <opentelemetry/sdk/trace/samplers/trace_id_ratio_factory.h>
+#include <opentelemetry/sdk/trace/tracer_provider.h>
+#include <opentelemetry/trace/provider.h>
+
+namespace google {
+namespace cloud {
+namespace otel {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+class BasicTracingConfigurationImpl : public BasicTracingConfiguration {
+ public:
+  explicit BasicTracingConfigurationImpl(
+      std::shared_ptr<opentelemetry::sdk::trace::TracerProvider> provider)
+      : provider_(std::move(provider)),
+        previous_(opentelemetry::trace::Provider::GetTracerProvider()) {
+    opentelemetry::trace::Provider::SetTracerProvider(
+        opentelemetry::nostd::shared_ptr<opentelemetry::trace::TracerProvider>(
+            provider_));
+  }
+
+  ~BasicTracingConfigurationImpl() override {
+    if (provider_) provider_->ForceFlush(std::chrono::microseconds(1000));
+    // Reset the global tracer provider.
+    opentelemetry::trace::Provider::SetTracerProvider(std::move(previous_));
+  }
+
+ private:
+  std::shared_ptr<opentelemetry::sdk::trace::TracerProvider> provider_;
+  opentelemetry::nostd::shared_ptr<opentelemetry::trace::TracerProvider>
+      previous_;
+};
+
+}  // namespace
+
+std::unique_ptr<BasicTracingConfiguration> ConfigureBasicTracing(
+    Project project, Options options) {
+  // Just return a nullptr if the project is not configured. This is intended
+  // as a function to make things easy, no reason to return complicated errors.
+  if (project.project_id().empty()) return {};
```

Comment:
I would remove this check. I do not think it makes it easier for customers to debug their configuration when they accidentally provide an empty project ID. And we don't want them to start relying on the `if (project.project_id().empty())` behavior.

If we care enough, we could `GCP_LOG(INFO) << "Empty project ID provided to tracing configuration.";`

Reply:
I think the check is a feature.  It let's me disable things easily.  OTOH, I don't care that much.  If we remove it, we need to return something sensible, either a `StatusOr<std::unique_ptr<...>>` or we add a `enabled()` accessor to the `BasicTracingConfiguration` class.  Your call.

Reply:
> I think the check is a feature

My intuition was that the check belongs in the application.

I think your intuition is that most applications will want to have the same code for tracing / non-tracing. And they would switch between the two with an env var. And it is more convenient if we write the `if` statement than them. Something like:

```cc
auto project_id = getenv("MY_PROJECT_ID");
ConfigureBasicTracing(gc::Project(project_id));
```

vs.

```cc
auto project_id = getenv("MY_PROJECT_ID");
if (project_id) {
  ConfigureBasicTracing(gc::Project(project_id));
}
```

I am happy to leave the code as is and go with your intuition.

> If we remove it, we need to return something sensible

(For the sake of arguing), a configuration with a useless exporter is a sensible return value.

GCP project IDs must be between 4 and 30 characters. But we do not also check that `project.project_id().size() < 4 || project.project_id().size() > 30`.